sympy/tensor/indexed.py

```diff
@@ -137,6 +140,24 @@ class Indexed(Expr):
     is_symbol = True
     is_Atom = True
 
+    @staticmethod
+    def _filter_assumptions(kw_args):
+        """Split the given dict into two parts: assumptions and not assumptions.
+           Keys are taken as assumptions if they correspond to a handler on ``Q``."""
+        assumptions = {k: v for k, v in kw_args.items() if k in _assume_defined}
+        Symbol._sanitize(assumptions)
+        # return assumptions, not assumptions
+        return assumptions, {k: v for k, v in kw_args.items() if k not in assumptions}
+
+    @staticmethod
+    def _set_assumptions(obj, assumptions):
+        """Set assumptions on obj, making sure to apply consistent values."""
+        tmp_asm_copy = assumptions.copy()
+        is_commutative = fuzzy_bool(assumptions.get('commutative', True))
+        assumptions['commutative'] = is_commutative
+        obj._assumptions = StdFactKB(assumptions)
+        obj._assumptions._generator = tmp_asm_copy  # Issue #8873
```

Comment:
this doesn't include the commutativity -- is that right?

Reply:
Not quite sure what you mean - what do you mean it doesn't include the commutativity?  The logic here is based on `Symbol`, which defaults to be commutative if not specified otherwise. This does the same thing, I believe.

Reply:
`assumptions` is potentially changed at line 157 but you store, in `_generator`, the copy, not the updated assumptions.

Reply:
But I confirm that this is the same as in Symbol.

Reply:
Aha, I see what you mean, yes. Seems to me that the copy is intentionally made before the change, so as to detect what was specified by the user vs. what the end result was. At least that's how I understand #8873.